cpp/src/slice2cpp/Gen.cpp

```diff
@@ -898,289 +874,184 @@ Slice::Gen::writeExtraHeaders(IceInternal::Output& out)
-        MetadataPtr meta = *q++;
-        string_view directive = meta->directive();
-        string_view arguments = meta->arguments();
+    // "cpp:no-default-include"
+    MetadataInfo noDefaultIncludeInfo = {
+        {&typeid(Unit)},
+        MetadataArgumentKind::NoArguments,
+    };
+    metadataInfo.emplace("cpp:no-default-include", std::move(noDefaultIncludeInfo));
 
-        // Issue friendly warning for cpp11 and cpp98 metadata what were removed as Slice does not issue warnings
-        // for unknown "top-level" metadata.
-        if (directive.find("cpp11") == 0 || directive.find("cpp98") == 0)
-        {
-            ostringstream ostr;
-            ostr << "ignoring invalid metadata '" << *meta << "'";
-            cont->unit()->warning(meta->file(), meta->line(), InvalidMetadata, ostr.str());
-            metadata.remove(meta);
-            continue;
-        }
+    // "cpp:no-stream"
+    MetadataInfo noStreamInfo = {
+        {&typeid(Unit)},
+        MetadataArgumentKind::NoArguments,
+    };
+    metadataInfo.emplace("cpp:no-stream", std::move(noStreamInfo));
 
-        if (directive.find("cpp:") != 0)
-        {
-            continue;
-        }
+    // "cpp:source-ext"
+    MetadataInfo sourceExtInfo = {
+        {&typeid(Unit)},
+        MetadataArgumentKind::SingleArgument,
+    };
+    metadataInfo.emplace("cpp:source-ext", std::move(sourceExtInfo));
 
-        if (operation && directive == "cpp:const" && arguments.empty())
-        {
-            continue;
-        }
+    // "cpp:source-include"
+    MetadataInfo sourceIncludeInfo = {
+        {&typeid(Unit)},
+        MetadataArgumentKind::RequiredTextArgument,
+    };
+    sourceIncludeInfo.mustBeUnique = false;
+    metadataInfo.emplace("cpp:source-include", std::move(sourceIncludeInfo));
 
-        if (directive == "cpp:type" && (arguments == "wstring" || arguments == "string"))
-        {
-            BuiltinPtr builtin = dynamic_pointer_cast<Builtin>(cont);
-            ModulePtr module = dynamic_pointer_cast<Module>(cont);
-            ClassDefPtr clss = dynamic_pointer_cast<ClassDef>(cont);
-            InterfaceDefPtr interface = dynamic_pointer_cast<InterfaceDef>(cont);
-            StructPtr strct = dynamic_pointer_cast<Struct>(cont);
-            ExceptionPtr exception = dynamic_pointer_cast<Exception>(cont);
-            if ((builtin && builtin->kind() == Builtin::KindString) || module || clss || strct || interface ||
-                exception)
+    // "cpp:unscoped"
+    MetadataInfo unscopedInfo = {
+        {&typeid(Enum)},
+        MetadataArgumentKind::NoArguments,
+    };
+    metadataInfo.emplace("cpp:unscoped", std::move(unscopedInfo));
+
+    // "cpp:view-type"
+    MetadataInfo viewTypeInfo = {
+        {&typeid(Sequence), &typeid(Dictionary)},
+        MetadataArgumentKind::RequiredTextArgument,
+        nullopt,
+        false,
+        true,
+    };
+    metadataInfo.emplace("cpp:view-type", std::move(viewTypeInfo));
+
+    // "cpp:type"
```

Comment:
The only "messy" thing in this PR is trying to validate the `cpp:type` metadata.
But it's more a product of the metadata being messy itself I think.

Reply:
I did spend some time thinking about it, but decided against adding any mechanisms to handle it because:

A) I believe that this is the _only_ metadata with two disjoint uses-cases like this.
So it's not worth complicating everything else to handle a single strange case.

B) I don't want to encourage this kind of metadata re-use. Realistically, these should just be two separate directives.
But they aren't, and we shouldn't break existing Slice definitions just to make the validation logic cleaner IMO.

The truly painful part is that the string flavors can be placed on containers, but the non-string flavors cannot.
Compare this to `cs:generic`. Yes `cs:generic` has a couple 'special' values like `cpp:type` does, but there's still consistent rules for where/when `cs:generic` can be used. Unlike `cpp:type` which has 2 separate rule sets.